android/app/src/main/java/com/example/moproapp/Circuit.kt

```diff
@@ -0,0 +1,52 @@
+package com.example.moproapp
+
+import CircuitInputs
+import android.content.Context
+import prepareCircuitInputs
+import readFileInChunks
+import uniffi.mopro.GenerateProofResult
+import uniffi.mopro.generateCircomProof
+import uniffi.mopro.verifyCircomProof
+
+
+class Circuit(zkeyFile:String, vkeyFile:String, _rapidsnarkInputs: CircuitInputs, _rawIntpus:MutableMap<String, List<String>>, context: Context) {
+
+    var zkeyPath:String = ""
+    var vkeyPath:String = ""
+    lateinit var rapidsnarkInputs: CircuitInputs
+    lateinit var rawInputs: MutableMap<String, List<String>>
+    // proof result
+    lateinit var rapidsnarkProof: ProveResponse
+    lateinit var arkworksProof: GenerateProofResult
+
+    init {
+        zkeyPath = getFilePathFromAssets(context,zkeyFile)
+        vkeyPath = getFilePathFromAssets(context,vkeyFile)
+        rapidsnarkInputs = _rapidsnarkInputs
+        rawInputs = _rawIntpus
+    }
+
+    fun proveRapidSnark() {
+        rapidsnarkProof = groth16Prove(
+            zkeyPath,
+            rapidsnarkInputs.wtnsBuffer
+        )
```

Comment:
actually you need to also include the witness generation part
```kotlin
        zkpTools.witnesscalcKeccak256_256_test(
            rapidsnarkInputs.circuitBuffer,
            rapidsnarkInputs.circuitSize,
            rapidsnarkInputs.jsonBuffer,
            rapidsnarkInputs.jsonSize,
            rapidsnarkInputs.wtnsBuffer,
            rapidsnarkInputs.wtnsSize,
            rapidsnarkInputs.errorMsg,
            rapidsnarkInputs.errorMsgMaxSize
        )
```

because the `generateCircomProof` also includes witness generation time

Reply:
or if you need to separate witness generation and proof generation in `generateCircomProof`
we need to figure out how to do it in `circom-prover`

previously I used a local crate and print out the result
or return the time
https://github.com/zkmopro/android-benchmark-app/blob/d1bce419af5d8b9c66e214c610f8afac526e4057/src/lib.rs#L34-L46

Reply:
"include the witness generation part" makes more sense to me. I think users only care about how long proof generation takes.

Reply:
fixed in [dac6de1](https://github.com/zkmopro/android-benchmark-app/pull/8/commits/dac6de1f72039df76e741c99877b233b6e5a0d8d)